### ISSUE TITLE
Introduce 'connect-certificate-hash' for WebTransport

### ISSUE THREAD

**Issue by jan-ivar** [w3c/webappsec-csp#683]
From https://github.com/w3c/webtransport/issues/59#issuecomment-2400880498: WebTransport acts like fetch wrt `connect-src: https://A.com` for normal use like:
```js
const wta = new WebTransport("https://A.com");
```
But the API also allows websites to connect using custom hashes:
```js 
const wtb = new WebTransport("https://B.com", {serverCertificateHashes: [{algorithm: "sha-256", value}]});
```
For this we think we need of a new CSP keyword, something like: `connect-certificate-hash: https://B.com/` e.g.
```js
Content-Security-Policy: connect-src https://A.com/;
                         connect-certificate-hash: https://B.com/
```
The idea is:
- if `connect-src` is absent, then custom certs are allowed
- if `connect-src` is present, `connect-certificate-hash` is needed to allow custom certs 

Does this make sense? Can someone help us with this?

**Comment by martinthomson** [comment]
See also https://github.com/JonathanLennox/raw-key-fingerprints

**Comment by dveditz** [comment]
I may be misreading your proposal. The text says "keyword", but the semi-colon after the a.com host (plus the later b.com host) made `connect-certificate-hash` into a separate "directive". An actual keyword like my `'unsafe-cert-hashes'` example above could work as an opt-in (note the single quotes that signify it is not a hostname). 

I'm also unclear on what threat this is trying to address. WebTransport requires script, and if you have malicious script running on your page then CSP has already failed you.

**Comment by vasilvv** [comment]
> Seems like either cert hashes are always safe, or they always require a site to opt-in whether or not it's using a CSP.

The issue we are trying to address here is that cert hashes change the assumptions under which the other constraints are interpreted.  For instance, if I say `connect-src https://example.com/`, in current world, this implies that an attacker would need a valid cert for `example.com` to exfiltrate data, whereas with cert hashes, the attacker can now exfiltrate data if it can perform a MITM on example.com.

**Comment by annevk** [comment]
https://github.com/w3c/webtransport/issues/623 is relevant here.

**Comment by jan-ivar** [comment]
Revised proposal:
```js
Content-Security-Policy: connect-src https://A.com/;
                         webtransport-certificate-hash: https://B.com/
```

**Comment by jan-ivar** [comment]
Revised proposal 2:
```js
Content-Security-Policy: connect-src 'certificate-hash' https://A.com https://B.com
```

Simplifies. When people add `connect-src` they limit by URL, which means no certificate hashes by default, unless they add the new `'certificate-hash'` keyword.

**Comment by mozfreddyb** [comment]
This CSP would allow all hashes, whatever the page author writes. What would it look like if the CSP were to only allow certain hashes that are known a priori?

**Comment by jan-ivar** [comment]
Given that hashes are meant to be short-lived, we're not sure CSP is a good place for them.

If the need this materializes in the future, we could consider add that then.

**Comment by jan-ivar** [comment]
Note that `connect-src` would not include `'certificate-hash'` by default.

**Comment by jan-ivar** [comment]
And to reiterate:
- if `connect-src` and `default-src` are absent (e.g. CSP not used), then hashes are allowed for backwards compatibility
- if `connect-src` or `default-src` are present, then hashes as DISALLOWED without the new keyword

IOW the weakening keyword ensures any URLs listed in CSP remain validated by both dns AND server certificates by default.

**Comment by dveditz** [comment]
You could imagine a policy that allows connections to a host by cert hash _instead of_ host name. This would rhyme with what `script-src` does with integrity hashes, but then the CSP folks would insist that those specific hashes be present in the policy. I'm sure that's what @mozfreddyb had in mind in his comment from last week. That may be something you add later if you need it.

As a strawman, and using `wthash-` as a placeholder prefix, it could work like this:
```http
Content-Security-Policy: connect-src a.com b.com wthash-xxxxxxxxxxxxxxxxxxxxx
```
The above could mean connections (fetch, web transport, web sockets) were allowed with valid certs to a.com and b.com, and web transport connections are allowed to any host that uses a certificate with the hash xxxxxxxxxxxxxxxx.

```http
Content-Security-Policy: connect-src a.com b.com wthash-xxxxxxxxxxxxxxxxxxxxx 'unsafe-webtransport-hashes'
```

The above could mean connections (fetch, web transport, web sockets) were allowed with valid certs to a.com and b.com, and web transport connections are allowed to any host with a certificate with the hash xxxxxxxxxxxxxxxxxx, OR web transport connections were allowed to a.com and b.com (ONLY) using ANY (unsafe) content-supplied hash.

An alternative interpretation would be to make `unsafe-webtransport-hashes` mutually exclusive with specified hashes. In that scenario the first example would mean that web transports is restricted to a.com and b.com like fetch and web sockets, but web transport could connect using a valid cert or the policy-allowed xxxxxxxxxxxxxxxxxxxxx. The policy author would have to choose between explicit ("safe") hashes and "unsafe" (unrestricted) hashes as in your example two comments up.  My second example would make no sense in this interpretation and we'd have to decide whether to ignore the redundant wthash, ignore the unsafe-hashes keyword, enforce `connect-src 'none';` to force authors to fix the error, or ignore the policy entirely (but fail open is bad!).

The second alternative doesn't fit as well with the rest of CSP and I much prefer the first, but I offer both to spark discussion. And for now you can ignore them both. Like the evolution of `script-src`, you could decide to add some variation on `wthash-` in a future version of the spec.

**Comment by martinthomson** [comment]
My understanding of the discussion at our recent meeting was that we wanted a general "hashes are OK", which would take the interpretation that `'unsafe-webtransport-hashes'` is a blanket approval for all hashes.  If we were to add wthash-XXX, which we decided could come later, then that would be a strict subset of `'unsafe-webtransport-hashes'`.  You would list specific hashes OR any hash with the keyword.  I think that's @dveditz's alternative.

The model that I'm using here is that when a connection is made and `connect-src` is consulted, that connection attempt is checked against the list of things in `connect-src` and only permitted if a match is found in the list.  @dveditz's second example doesn't fail closed under that model, so it makes a kind of sense, but you wouldn't want to use that construction, because the wthash-XXX is redundant.

> web transport connections were allowed to a.com and b.com (ONLY) using ANY (unsafe) content-supplied hash.

I think that there is possibly some misunderstanding here.  If something uses a hash, that is not connecting to a name, it is connecting to a hash.  So while we might need a name for DNS lookups, that is not the authority in question.  The authority is derived from the hash, so it doesn't make sense to say "connect to a.com using a content-supplied hash", even if that is sort of how the API is spelled out.  Logically, when a hash is supplied, THAT is the operative origin (even though content delivered using webtransport, like websockets, is largely outside of our typical model, or at least it is assumed to be readable to the origin it is delivered to).  At least, that is my understanding of the situation.

**Comment by dveditz** [comment]
At the WebAppSec WG June 18 meeting we took up this topic and approved adding a `'unsafe-webtransport-hashes'` to CSP's `connect-src` directive (which could also be specified in the `default-src` directive if the policy had no `connect-src` directive).

@jan-ivar would you like to take a stab at making a PR for this? It would be something sort of like PR #457 for webrtc where you'd describe what the directive does in the connect-src section, and then define an algorithmic hook that the WebTransport spec calls. It would be something like https://w3c.github.io/webappsec-csp/#connect-src-pre-request that the Fetch spec calls, but it probably makes more sense to have a separate one rather than complicate the existing one with checks whether it's being called from Fetch or called from Web Transport.

**Comment by quantumpacket** [comment]
> At the WebAppSec WG June 18 meeting we took up this topic and approved adding a 'unsafe-webtransport-hashes' to CSP's connect-src directive (which could also be specified in the default-src directive if the policy had no connect-src directive).

Even if `connect-src` is being defined, can `unsafe-webtransport-hashes` still be used in `default-src` where it will be applied to it? Or does it have to be used in `connect-src` if present?

**Comment by antosart** [comment]
> Even if connect-src is being defined, can unsafe-webtransport-hashes still be used in default-src where it will be applied to it? Or does it have to be used in connect-src if present?

No, it has to be in `connect-src`. For how CSP works, more specific directives override fallback directives. Meaning that if `connect-src` is present, `default-src` will not be taken into consideration for all types of resource requests which are governed by `connect-src` (and in particular for webtransport connections).
